tree/ntuple/v7/src/RNTuple.cxx

```diff
@@ -69,15 +69,15 @@ void ROOT::Experimental::RNTupleReader::ConnectModel(RNTupleModel &model)
 {
    auto &fieldZero = model.GetFieldZero();
    // We must not use the descriptor guard to prevent recursive locking in field.ConnectPageSource
-   fieldZero.SetOnDiskId(fSource->GetSharedDescriptorGuard()->GetFieldZeroId());
-   for (auto &field : fieldZero) {
-      // If the model has been created from the descritor, the on-disk IDs are already set.
+   DescriptorId_t fieldZeroId = fSource->GetSharedDescriptorGuard()->GetFieldZeroId();
+   fieldZero.SetOnDiskId(fieldZeroId);
+   for (auto &field : fieldZero.GetSubFields()) {
```

Comment:
It took me a while and a discussion with Vincenzo to understand that before
```c++
for (auto &field : fieldZero) {
```
iterated over the entire tree of subfields while the new
```c++
for (auto &field : fieldZero.GetSubFields()) {
```
only looks at the direct subfields and the recursion is in `RFieldBase::ConnectPageSource` (basically I was trying to understand how it worked before and failed to see the recursion that is in `RSchemaIterator`). Maybe you can add a comment to help the future confused reader, something like
```c++
// Iterate only over fieldZero's direct subfields; their descendants are recursively handled in RFieldBase::ConnectPageSource
```